### PR TITLE
Enforce .Elem.Fields (instead of .Fields) for object types

### PR DESCRIPTION
**Pull request by iwahbe**
This is a reversal of our previous stance that .Fields should be used. This aligns the
fields argument with [schema generation](https://github.com/pulumi/pulumi-terraform-bridge/blob/72b0b7d33b6d9f0dce7c09b31951ade3711a025d/pkg/tfgen/generate.go#L395-L398).

We *need* to choose `.Elem.Fields` as our 1 representation (instead of `.Fields`) because
of an ambiguity in `shim.Schema`. From a comment in the PR:

> To prevent confusion, users are barred from specifying
> information on ps directly, they must set .Fields on ps.Elem:
> ps.Elem.Fields.
>
> We need to make this choice (instead of having users set
> information on .Fields (and forbidding ps.Elem.Fields) because
> the [shim.Schema] representation is not capable of
> distinguishing between Map[Object] and Object. SDKv{1,2}
> providers are not able to express Map[Object], but Plugin
> Framework based providers are.

While this is technically a breaking change, I don't expect users to be broken. Any user
that would break on this change would have been unable to addapt the last several bridge
versions (for example: https://github.com/pulumiverse/pulumi-talos/pull/99).

Fixes https://github.com/pulumi/pulumi-terraform-bridge/issues/2185

I have tested this change on talos, AWS, GCP and Azure and it doesn't trigger an error.

### CODE REVIEW

**Review by VenelinMartinov — CHANGES_REQUESTED** [review]
Can we add some tfgen type override integration tests? We need one for 2185 as well as some tests around Set, List, Object, Map[Object], Map etc.

File: pkg/tfgen/generate.go [hunk 1]
```diff
@@ -398,16 +398,10 @@ func (g *Generator) makePropertyType(typePath paths.TypePath,
...
-	case shim.TypeList, shim.TypeSet:
-		if tfbridge.IsMaxItemsOne(sch, info) {
-			flatten = true
-		}
-	}
+	flatten := tfbridge.IsMaxItemsOne(sch, info)
 
-	// The remaining cases are collections, List[T], Set[T] or Map[T], and recursion needs NewElementPath except for
-	// flattening that stays at the current path.
+	// The remaining cases are collections, List[T], Set[T] or Map[T], and recursion needs NewElementPath
```
[the remark this answers is not in the document]
Comment (iwahbe) [reply pkg/tfgen/generate.go:403]: Pulled into https://github.com/pulumi/pulumi-terraform-bridge/pull/2313.

**Comment by iwahbe** [comment]
> Can we add some tfgen type override integration tests? We need one for 2185 as well as some tests around Set, List, Object, Map[Object], Map etc.

I have added integration tests. See [Add integration tests](https://github.com/pulumi/pulumi-terraform-bridge/pull/2309/commits/eaa7cdf13a1a62517ce39c1943f3f6caf8c5c159).

File: pf/tfgen/tfgen_test.go [hunk 2]
```diff
@@ -90,6 +96,374 @@ func (*schemaTestProvider) DataSources(ctx context.Context) []func() datasource.
...
+						},
+					},
+				},
+			},
+			info: &tfbridge.ResourceInfo{Fields: map[string]*tfbridge.SchemaInfo{
+				"a1": {Fields: map[string]*tfbridge.SchemaInfo{
+					"invalid": {},
+				}},
+			}},
+			expectedError: autogold.Expect("test_res: [{a1}]: .Fields should .Elem.Fields"),
```
Comment (VenelinMartinov) [inline pf/tfgen/tfgen_test.go:255]: typo

File: pf/tfgen/tfgen_test.go [hunk 3]
```diff
@@ -90,6 +96,374 @@ func (*schemaTestProvider) DataSources(ctx context.Context) []func() datasource.
...
+			if tt.expectedError != nil {
+				require.Error(t, err)
+				tt.expectedError.Equal(t, err.Error())
+				return
+			}
+			require.NoError(t, err)
+			var b bytes.Buffer
+			require.NoError(t, json.Indent(&b, res.ProviderMetadata.PackageSchema, "", "    "))
+			autogold.ExpectFile(t, autogold.Raw(string(
+				bytes.ReplaceAll(b.Bytes(), []byte("\r\n"), []byte{'\n'}),
```
Comment (VenelinMartinov) [inline pf/tfgen/tfgen_test.go]: Do we need the carriage return replacement if we are not running on windows?

File: pkg/tfbridge/info/validate.go [hunk 4]
```diff
@@ -107,48 +106,61 @@ func (c *infoCheck) checkProperty(path walk.SchemaPath, tfs shim.Schema, ps *Sch
...
+		if len(ps.Fields) > 0 {
+			c.error(path, errCannotSpecifyFieldsOnListOrSet)
+		}
+
 		c.checkElem(path.Element(), elem, ps.Elem)
 
+		switch tfs.Type() {
+		case shim.TypeSet, shim.TypeList:
+		default:
+			if ps.MaxItemsOne != nil {
```
Comment (VenelinMartinov) [inline pkg/tfbridge/info/validate.go:124]: nit: `if tfs.Type() != Set && != List` would read better here.

File: pkg/tfbridge/info/validate.go [hunk 5]
```diff
@@ -107,48 +106,61 @@ func (c *infoCheck) checkProperty(path walk.SchemaPath, tfs shim.Schema, ps *Sch
...
-	}
-
 	switch elem := tfs.Elem().(type) {
 
 	// s.Elem() case 1
 	//
 	// There is a simple element type here, so we just recursively validate that.
 	case shim.Schema:
+		if len(ps.Fields) > 0 {
+			c.error(path, errCannotSpecifyFieldsOnListOrSet)
```
Comment (VenelinMartinov) [inline pkg/tfbridge/info/validate.go:116]: This might be a map - the error here is incorrect.

File: pf/tfgen/tfgen_test.go [hunk 6]
```diff
@@ -90,6 +97,376 @@ func (*schemaTestProvider) DataSources(ctx context.Context) []func() datasource.
...
+				},
+			},
+			info: &tfbridge.ResourceInfo{
+				Fields: map[string]*tfbridge.SchemaInfo{
+					"a1": {Fields: map[string]*tfbridge.SchemaInfo{
+						"invalid": {},
+					}},
+				},
+			},
+			expectedError: autogold.Expect("test_res: [{a1}]: cannot specify .Fields on a List[T] or Set[T] type"),
```
Comment (VenelinMartinov) [inline pf/tfgen/tfgen_test.go:318]: The error here is misleading - we have a map type, not a list or set.

File: pf/tfgen/tfgen_test.go [hunk 7]
```diff
@@ -90,6 +97,376 @@ func (*schemaTestProvider) DataSources(ctx context.Context) []func() datasource.
...
+			info: &tfbridge.ResourceInfo{Fields: map[string]*tfbridge.SchemaInfo{
+				"a1": {Elem: &tfbridge.SchemaInfo{
+					Fields: map[string]*tfbridge.SchemaInfo{
+						"n1": {Type: "number"},
+					},
+				}},
+			}},
+		},
+		{
+			// This test case reproduces https://github.com/pulumi/pulumi-terraform-bridge/issues/2185
```
Comment (VenelinMartinov) [inline pf/tfgen/tfgen_test.go:194]: thanks!

File: pkg/tfgen/generate_schema_test.go [hunk 8]
```diff
@@ -189,7 +189,7 @@ func TestNestedDescriptions(t *testing.T) {
 	schema, err := GenerateSchema(provider, diag.DefaultSink(io.Discard, io.Discard, diag.FormatOptions{
 		Color: colors.Never,
 	}))
-	assert.NoError(t, err)
+	require.NoError(t, err)
```
Comment (VenelinMartinov) [inline pkg/tfgen/generate_schema_test.go:192]: Is there a reason to prefer `require`?

**Review by VenelinMartinov — CHANGES_REQUESTED** [review]
thanks for adding the tests!

File: pkg/tfgen/generate_schema_test.go (hunk 8 above)
Comment (iwahbe) [reply pkg/tfgen/generate_schema_test.go:192]: Yes. `require` stops the test when it errors, while `assert` just fails the test. 

If `err != nil`, then we don't want to assert that the JSON is equal (we know it won't be), so we should stop the test (with `require`).

File: pkg/tfbridge/info/validate.go [hunk 9]
```diff
@@ -107,48 +106,56 @@ func (c *infoCheck) checkProperty(path walk.SchemaPath, tfs shim.Schema, ps *Sch
...
```
Comment (VenelinMartinov) [inline pkg/tfbridge/info/validate.go:116]: Can we rename the variable to `CannotSpeciftFieldsOnCollection` - it has kept the old name?

File: pkg/tfbridge/info/validate.go [hunk 10]
```diff
@@ -107,48 +106,56 @@ func (c *infoCheck) checkProperty(path walk.SchemaPath, tfs shim.Schema, ps *Sch
...
+		}
+
+		if t := tfs.Type(); t != shim.TypeSet && t != shim.TypeList && ps.MaxItemsOne != nil {
```
Comment (VenelinMartinov) [inline pkg/tfbridge/info/validate.go:119]: Isn't this check essentially the same as the one under .Resource?

**Review by VenelinMartinov — APPROVED** [review]
looks good - two last small comments

File: pkg/tfbridge/info/validate.go [hunk 11]
```diff
@@ -141,14 +116,52 @@ func (c *infoCheck) checkProperty(path walk.SchemaPath, tfs shim.Schema, ps *Sch
...
+			// the [shim.Schema] representation is not capable of
+			// distinguishing between Map[Object] and Object. SDKv{1,2}
+			// providers are not able to express Map[Object], but Plugin
+			// Framework based providers are.
+			if len(ps.Fields) > 0 {
+				c.error(path, errFieldsShouldBeOnElem)
+			}
+			fallthrough
+		default:
+			if ps.MaxItemsOne != nil {
```
Comment (VenelinMartinov) [inline pkg/tfbridge/info/validate.go:161]: This is now redundant, right, as we check for non-valid max items one above.

**Pull request merged**